### PR TITLE
feat: plan agent refinement, feature discovery

### LINKED ISSUE #509: Auto-generate plan before multi-step execution
## Observation

During multi-step data diff and migration validation tasks (comparing `customer`, `orders`, and `lineitem` tables across `pg_source` and `pg_target`), no todo plan was generated upfront before execution began. The plan only appeared mid-way or at the end — after most of the work had already run. The user had to explicitly point this out before the behaviour was corrected.

## Improvement Request

For any multi-step execution process, the agent should:

1. **Show the plan first** — before running a single query or tool call, present the full step-by-step plan to the user (e.g. schema inspection → row counts → missing rows → value diff → financial impact → write report).

2. **Respect user approval preference** — support two modes:
   - **Auto-approved:** plan is shown and execution begins immediately
   - **Manually approved:** plan is shown and the agent waits for explicit user confirmation before proceeding

3. **Live status updates** — as each step executes, update the plan in real time (`pending` → `in_progress` → `completed`), giving the user full visibility into progress.

4. **Use conversation context for feedback** — when submitting feedback, pull in relevant session details (queries run, findings, corrections made) as supporting context.

## Session Context

This was observed across three data comparison prompts:
- Full migration validation: `customer`, `orders`, `lineitem` on pg_source vs pg_target
- Same-query baseline check: Q1 1994 lineitem on pg_source
- Targeted cross-database diff: Q1 1994 lineitem pg_source vs pg_target, checking `l_extendedprice`

In all three cases, the 6-phase validation workflow (schema → row counts → missing rows → value diff → financial impact → report) ran without a plan being shown first. The agent only started tracking phases after the user explicitly flagged the missing behaviour mid-session.

## Example Tasks Where This Matters

- Data diff validations across databases
- Migration completeness checks (multi-table)
- Any workflow with 3+ sequential tool calls
- ETL regression and query refactor verification

---

### Metadata

| Field | Value |
|-------|-------|
| CLI Version | 0.0.0-feat/data-parity-skill-improvements-202603270807 |
| Platform | darwin |
| Architecture | arm64 |
| OS Release | 25.3.0 |
| Category | improvement |
| Working Directory | altimate-data-parity-demo |
| Session ID | ses_2d19b4e60ffeotjlpymni4XuIN |

### PR DESCRIPTION
**Pull request by anandgupta42**
## Summary

Driven by telemetry analysis of altimate-code's first launch week, this PR addresses two P1 issues:

- **Plan agent completion rate at 64%** — users abandon plans and retry 2-3 times because there's no way to refine without restarting. Plan telemetry was also entirely null (couldn't measure improvements).
- **Feature discovery drop-off** — 53% of users connected a warehouse but only 19% found schema tools, 14% dbt, 5% lineage. Users didn't know what to do after connecting.

**Closes #509** — Auto-generate plan before multi-step execution (two-step outline-first approach)
**Partially addresses #355** — Plan mode UX improvements (refinement loop, approval detection)
**Partially addresses #206** — Skill discovery (dbt auto-detection now recommends skills proactively)

---

### Critical paths to review carefully

1. **`src/session/prompt.ts` (lines 623-694)** — Plan refinement detection. This runs on every step of the plan agent loop. Verify:
   - The `agent.name === "plan"` guard ensures non-plan agents are never affected
   - `planHasWritten` is only set via `Filesystem.exists()` (no false positives)
   - The revision cap at 5 injects a synthetic message but does NOT persist to DB
   - Phrase classification: rejection > refinement qualifier > approval > refine (default)

2. **`src/session/prompt.ts` (line 369)** — `sessionAgentName` fix. Was previously set only inside `if (step === 1)`, now set unconditionally from `lastUser.agent`. Verify this doesn't break other agents.

3. **`src/altimate/tools/warehouse-add.ts` (lines 47-98)** — Post-connect suggestions use `Promise.race` with 1.5s timeout. Verify the timeout doesn't cause unhandled promise rejections.

### Safe to skim (low risk)

- `src/altimate/tools/post-connect-suggestions.ts` — New file, no existing code modified
- `src/altimate/tools/sql-execute.ts`, `sql-analyze.ts`, `schema-inspect.ts`, `schema-index.ts` — Each appends ~3 lines of suggestion text, all wrapped in try/catch
- `src/session/prompt/plan.txt` — Additive prompt instructions, no deletions
- `src/altimate/telemetry/index.ts` — Type additions only (new event types)
- `src/tool/skill.ts` — One-line change: adds `trigger` field to existing telemetry call

### How to test locally

```bash
cd packages/opencode

# Run all 246 tests (each file individually to avoid cross-file module loading)
for f in test/session/plan-layer-e2e.test.ts test/altimate/feature-discovery-e2e.test.ts test/altimate/performance-regression.test.ts test/altimate/plan-refinement.test.ts test/altimate/post-connect-suggestions.test.ts test/telemetry/plan-skill-telemetry.test.ts test/telemetry/telemetry.test.ts; do
  bun test "$f" --timeout 30000
done

# Typecheck
bun turbo typecheck

# Run from source to manually test plan agent
bun run --conditions=browser ./src/index.ts
```

### Questions for reviewer
- Is the 1.5s timeout for warehouse-add suggestions aggressive enough? Could lower to 500ms if latency is a concern.
- Should the plan revision cap (5) be configurable, or is a hardcoded limit fine for now?
- The `refinementQualifiers` list (" but ", " however ", " except ", etc.) — are there common phrases we're missing?

---

## Plan Agent Improvements
- Two-step plan approach: 3-5 bullet outline first → user confirms → expand to full plan
- Plan refinement loop: users can edit/iterate instead of restarting (capped at 5 revisions)
- Smart phrase classification: approval ("looks good"), rejection ("no", "stop"), and refinement ("yes, but change X") with qualifier overrides and word-boundary matching
- Revision cap communication: when limit reached, synthetic message tells the LLM to inform the user
- **Bug fix:** `agent_outcome` telemetry for plan sessions was emitting null for `duration_ms`, `tool_calls`, `generations` — root cause was `sessionAgentName` assignment too late in the loop

## Feature Discovery & Progressive Disclosure
- Post-warehouse-connect contextual suggestions (schema index, SQL analysis, lineage, PII audit)
- Progressive tool suggestions: `sql_execute` → `sql_analyze` → `schema_inspect` → `lineage_check`
- **Deduplication**: each progressive hint shown at most once per session (no repetitive tips)
- dbt project auto-detection → recommends `/dbt-develop`, `/dbt-troubleshoot`, `/dbt-analyze` skills
- All suggestions wrapped in try/catch — never block core flow

## Performance & UX
- Warehouse-add suggestions use `Promise.all` + `Promise.race` with 1.5s timeout — slow schema/dbt checks silently skipped instead of blocking response
- Progressive suggestion generation: < 0.05ms per call (verified via benchmarks)
- Phrase detection: < 0.002ms per call (100k iterations in < 200ms)
- Suggestion dedup via `shownProgressiveSuggestions` Set — zero allocation on repeated calls

## Telemetry Instrumentation
- New `plan_revision` event: tracks revision count and action (`refine` | `approve` | `reject` | `cap_reached`)
- New `feature_suggestion` event: tracks which suggestions shown, type, warehouse
- Added `trigger` field to `skill_used` event: `user_command` | `llm_selected` | `auto_suggested`
- `classifySkillTrigger()` helper function
- Fixed `sessionAgentName` so `agent_outcome` always has correct agent name

### Unit & Integration Tests (246 total, all passing)
- [x] `telemetry.test.ts` — 90 tests: full telemetry regression suite with `ALL_EVENT_TYPES` constant
- [x] `plan-layer-e2e.test.ts` — 68 tests: **plan layer safety** — phrase classification (42 edge cases including unicode, multiline, adversarial 70k input), non-plan agent safety (3), sessionAgentName fix (2), revision cap (4), adversarial (4), import safety (2)
- [x] `feature-discovery-e2e.test.ts` — 29 tests: full warehouse→suggestions flow with mocked Dispatcher, progressive chain, plan refinement session, telemetry events
- [x] `post-connect-suggestions.test.ts` — 20 tests: contextual suggestions, progressive disclosure, dedup, telemetry
- [x] `plan-skill-telemetry.test.ts` — 14 tests: trigger classification, event types, runtime `Telemetry.track()` verification
- [x] `plan-refinement.test.ts` — 13 tests: revision counter, cap, approval detection, telemetry
- [x] `performance-regression.test.ts` — 12 tests: suggestion gen < 50ms/1000x, lookup < 50ms/10kx, phrase detection < 200ms/100kx, output determinism

### CI & Type Safety
- [x] Typecheck passes (`bun turbo typecheck` — 5/5 packages)
- [x] No new test failures (pre-existing in `regression.test.ts` — unrelated `Telemetry.shutdown` issue)

### Post-Deploy Verification
- [ ] Verify plan agent telemetry populates in App Insights after deploy
- [ ] Monitor `feature_suggestion` events to measure warehouse→schema conversion
- [ ] Track plan completion rate (target: 64% → 80%)
- [ ] Verify suggestion dedup works in production (no repeated tips)

### Target Metrics (4-week)
| Metric | Current | Target |
|--------|---------|--------|
| Plan completion rate | 64% | 80% |
| Plan abandon-then-retry rate | ~50% | < 20% |
| Warehouse → Schema conversion | 36% | 55% |
| Warehouse → dbt conversion | 24% | 40% |
| Skill trigger tracking | 0% | 100% |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Contextual post-connection suggestions tailored to warehouse type, schema/indexing status, DBT detection, and connection count.
  * Progressive-disclosure tips that surface related tools as you use the app.
  * Two‑step plan workflow: short outline + user confirmation, with tracked plan revisions and a revision cap.

* **Improvements**
  * Telemetry expanded to classify skill triggers and record new events for plan revisions and feature suggestions.

* **Documentation**
  * Added docs describing post-connection suggestions and updated Plan-mode behavior.

* **Tests**
  * Extensive unit, integration, and E2E tests covering suggestions, progressive disclosure, plan revisions, and telemetry.

### CODE REVIEW

File: packages/opencode/src/altimate/tools/warehouse-add.ts [hunk 1]
```diff
@@ -41,10 +44,63 @@ IMPORTANT: For private key file paths, always use "private_key_path" (not "priva
...
+            if (suggestionCtx.dbtDetected) suggestionsShown.push("dbt-develop", "dbt-troubleshoot")
+            if (suggestionCtx.connectionCount > 1) suggestionsShown.push("data_diff")
+            PostConnectSuggestions.trackSuggestions({
+              suggestionType: "post_warehouse_connect",
+              suggestionsShown,
+              warehouseType: result.type,
+            })
+          }
+        } catch {
+          // Suggestions must never break the add flow
```
Comment (suryaiyer95) [inline packages/opencode/src/altimate/tools/warehouse-add.ts]: [suggestion](https://pullpo.io/cc?l=suggestion) 
Can add telemetry on these failures?

File: packages/opencode/src/session/prompt.ts [hunk 2]
```diff
@@ -610,6 +620,79 @@ export namespace SessionPrompt {
...
+            const approvalPhrases = ["looks good", "proceed", "approved", "approve", "lgtm", "go ahead", "ship it", "yes", "perfect"]
+
+            const isRejectionPhrase = rejectionPhrases.some((phrase) => userText.includes(phrase))
+            const isRejectionWord = rejectionWords.some((word) => {
+              const regex = new RegExp(`\\b${word}\\b`)
+              return regex.test(userText)
+            })
+            const isRejection = isRejectionPhrase || isRejectionWord
+            const isApproval = !isRejection && !hasRefinementQualifier && approvalPhrases.some((phrase) => userText.includes(phrase))
+            const action = isRejection ? "reject" : isApproval ? "approve" : "refine"
```
[the remark this answers is not in the document]
Comment (anandgupta42) [reply packages/opencode/src/session/prompt.ts]: Good question — right now all three actions (approve/reject/refine) are **telemetry-only**. The plan agent continues in plan mode regardless of classification. The session doesn't auto-switch to builder on approval.

This is intentional for this PR — the classification gives us data to understand user intent (e.g., how often users approve vs iterate), which we'll use to design the auto-switch behavior in a follow-up (partially addresses #355).

In a future PR we can use `action === "approve"` to trigger an actual mode transition to builder with the plan as context.

File: packages/opencode/src/altimate/tools/warehouse-add.ts (hunk 1 above)
Comment (anandgupta42) [reply packages/opencode/src/altimate/tools/warehouse-add.ts]: Good catch — added `core_failure` telemetry with `error_class: "suggestion_failure"` in the catch block. This way we can track how often suggestions fail and what errors occur, without breaking the warehouse add flow.

Pushed in the latest commit.

**Review by mdesmet — APPROVED** [review]
LGTM If we also could see the full plan as md file would also be great, that is easier to read than on CLI in IDE.

**Comment by anandgupta42** [comment]
Good question — the 5-revision cap is **per session**, not per day. It counts every user message after the initial plan is written, regardless of whether it's a major restructuring or a minor tweak.

The counter (`planRevisionCount`) resets when a new session starts. So a user can:
- Session 1: refine plan 5 times → cap reached → start new session
- Session 2: get 5 fresh revisions

When the cap is hit, the LLM is told via a synthetic system message to inform the user and suggest either finalizing the plan or starting a new planning session. The `cap_reached` action is also tracked in telemetry so we can see how often users hit it and adjust the limit if needed.

If 5 turns out to be too restrictive based on telemetry data, we can make it configurable later.
